shipment_advice/models/shipment_advice.py

```diff
@@ -295,64 +304,96 @@ def _lock_records(self, records):
-                        wiz.with_context(
-                            button_validate_picking_ids=picking.ids
-                        ).process()
-                    else:
-                        picking._action_done()
-            else:
-                backorder_policy = (
-                    shipment.company_id.shipment_advice_outgoing_backorder_policy
-                )
-                self._lock_records(self.loaded_picking_ids)
-                if backorder_policy == "create_backorder":
-                    for picking in self.loaded_picking_ids:
-                        if picking.state in ("cancel", "done"):
-                            continue
-                        if picking._check_backorder():
-                            wiz = wiz_model.create({})
-                            wiz.pick_ids = picking
-                            wiz.with_context(
-                                button_validate_picking_ids=picking.ids
-                            ).process()
-                        else:
-                            picking._action_done()
-                else:
-                    for picking in self.loaded_picking_ids:
-                        if picking.state in ("cancel", "done"):
-                            continue
-                        if not picking._check_backorder():
-                            # no backorder needed means that all qty_done are
-                            # set to fullfill the need => validate
-                            picking._action_done()
-                # Unplan moves that were not loaded and validated
-                moves_to_unplan = (
-                    self.loaded_move_line_ids.move_id | self.planned_move_ids
-                ).filtered(
-                    lambda m: m.state not in ("cancel", "done") and not m.quantity_done
-                )
-                moves_to_unplan.shipment_advice_id = False
-            if not shipment.departure_date:
-                shipment.departure_date = fields.Datetime.now()
-            shipment.state = "done"
-        return True
+
+    def _validate_picking(self, picking, backorder_policy="create_backorder"):
+        self.ensure_one()
+        self._lock_records(picking)
+        try:
+            with self.env.cr.savepoint():
+                if (
+                    picking._check_backorder()
+                    and backorder_policy == "create_backorder"
+                ):
+                    wiz = self.env["stock.backorder.confirmation"].create({})
+                    wiz.pick_ids = picking
+                    wiz.with_context(button_validate_picking_ids=picking.ids).process()
+                elif not picking._check_backorder():
+                    picking._action_done()
+        except UserError as error:
+            self.write(
+                {
+                    "state": "error",
+                    "error_message": self._get_error_message(error, picking),
+                }
+            )
```

Comment:
The method will be called for each picking in a dedicated job but every time a validation fail it will overwrite the previous error message that could have been set by an other picking validation. 

The validation should probably be sequential to avoid concurrent update on quant update when a move is validated for the same product in the same place concurrently. In such a case you could store every error into a json field to keep track of every errors encountered by the validation process. The error message could be computed from the json field and the json field reset every time the action done is launched. This works only if the picking validation is sequential to avoid concurrent update on the same `shipment.advice` record by concurrent failing picking validation jobs.

Reply:
> The method will be called for each picking in a dedicated job but every time a validation fail it will overwrite the previous error message that could have been set by an other picking validation.
> 

Tracking is enabled on the erreor message

> The validation should probably be sequential to avoid concurrent update on quant update when a move is validated for the same product in the same place concurrently. In such a case you could store every error into a json field to keep track of every errors encountered by the validation process. The error message could be computed from the json field and the json field reset every time the action done is launched. This works only if the picking validation is sequential to avoid concurrent update on the same `shipment.advice` record by concurrent failing picking validation jobs.

The validation fails remain exceptions, and the normal case is the success of the job. 

In this logic, it is preferable, IMO, to keep the validations run in parallel and let the queue_job handle rollbacks and retries.

In case of concurrency, eventually one of the jobs will succeed, and the others will be rescheduled. This way, we maintain a relatively fast validation process.

In the worst-case scenario, the administrator can set the channel capacity to one to run jobs sequentially.